enhancements/machine-config/manage-boot-images.md

```diff
@@ -319,6 +324,194 @@ When [MachineDeployments](https://cluster-api.sigs.k8s.io/developer/architecture
+
+**AWS Streams:**
+- IPI - Standard installer-provisioned infrastructure on AWS
+- Marketplace - AWS Marketplace published images
+- ROSA - Red Hat OpenShift Service on AWS
+
+**GCP Streams:**
+- IPI - Standard installer-provisioned infrastructure on GCP
+- Marketplace - GCP Marketplace published images (OCP, OPP, OKE variants)
+
+**Azure Streams:**
+- IPI/ARO - Standard installations and Azure Red Hat OpenShift (both HyperV Gen1 and Gen2)
+- Marketplace - Azure Marketplace published images (paid offerings with OCP, OPP, OKE variants)
+
+**OKD Streams:**
+- SCOS-OKD - Singular SCOS stream on AWS
+
+The MCO examines the current MachineSet boot image, determines its stream, and then determines the desired image from the appropriate stream metadata. Once all known streams are handled, the MCO should enter a degraded state if it expects to manage a boot image but cannot determine its stream.
+
+#### Platform-Specific Stream Detection and Update Strategy
+
+##### AWS
+
+###### History Bootimage Tracking (Existing Clusters)
+
+For existing upgrade clusters, determining the boot image source requires examining multiple data sources:
+
+**AWS MachineSet backed installer-shipped image:**
+- Create a historical list of all installer-shipped AMIs from RHCOS.json files
+- Match current MachineSet AMI against this historical data
+
+**ROSA:**
+- Create a list of all ROSA AMIs based on GitLab cluster image set metadata (see Red Hat internal Gitlab instance - service/clusterimagesets)
+
+**AWS Marketplace:**
+Two potential approaches:
+1. Filter out all non-ROSA marketplace published images and curate a list
+2. Default to not supported - require user to update to a "new" marketplace image once skew enforcement is in place
+
+**Detection fallback:**
+If the boot image is not found in any historical list:
+- Use AWS SDK to check publisher and RHCOS version:
+  - **Deregistered/Not Found**: We will estimate the skew based on the install version of the cluster, and have the skew enforcement rules apply to it as normal
+  - **Marketplace-published**: Could be UPI or ROSA - apply normal skew enforcement rules, but require update to future images with proper metadata
+  - **RHCOS**: Follow regular skew rules (consider raising warning since it should be in the historical list)
+
+###### Future Bootimage Tracking (New Images)
+
+**Installer changes:**
+
+Create multiple JSON files of AMI variant metadata for different streams:
+- rhcos (standard IPI)
+- rosa
+- marketplace-ocp
+- marketplace-oke
+- marketplace-opp
+
+**Image creation:**
+
+- All variants are tagged with `variantType` metadata to enable deterministic stream detection
+- The tagging will be done with AWS's [tagSet](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Image.html) field, with `variantType as the Key and stream name as the value
```

Comment:
Any suggestions for the images already present in AWS? Should we somehow tag them as well?

Reply:
Currently we don't plan on tagging existing images. Parsing of existing images that are referenced will be a mix between hard-coded lists we have today of images we know we've shipped, as well as the publisher metadata (see History Bootimage Tracking (Existing Clusters) section above).

If for some reason we don't know where an image came from, we hope to do that as a one off thing and ask the customer to take some manual action to update to the right image for their cluster.